docs/internal_abi.md

```diff
@@ -0,0 +1,323 @@
+
+If the payload is an integer number of bytes, the payload is stored first (independent of byte order) so that the payload can benefit from strong alignment, which avoids the payload accidentally spanning a cache line.
+In this case, all elements of the payload are stored starting from the beginning of the payload.
+
+*Rationale:* The payload can be loaded with a single aligned load on any architecture and depending on the target byte order then needs to do either a single zero-extending shift right or a left shift followed by a zero-extending right shift.
+
+If storing the discriminator bits at either end of the space results in different numbers of complete bytes for the payload then the discriminator should stored at the end that results in the larger number of complete bytes for the payload.
+If there are no discriminator bits, this rule is applied as if there were a single bit.
+Payload values that are smaller than the payload space are then stored aligned to the opposite end of the value to the discriminator.
+
+*Rationale:* This case is reached if we are packing a union type inside the significand bits of a signalling NaN and guarantees that we can load the payload in a single aligned load and need to zero bits at only one end to extract the payload.
+
+In other cases, the discriminator bits are stored in the low bits (as defined by the platform byte order) of the object and the payload in the high bits.
+
+*Rationale:* The discriminator is easy to mask off with a right shift and extract with an xor with an immediate operand for comparisons.
+
+The types identified in a discriminator are sorted lexically and discriminator values are assigned in order.
+Singletons are sorted before all other types.
+
+*Rationale:* The representation of `A | B` and `B | A` should be the same.
+It is useful to sort singletons before pointers so that the common case of `T* | NotAThing` is able to use null-pointer checks and branches, which are typically optimised aggressively in modern microarchitectures.
+
+{*Note:* I'm not sure lexical sorting will work because types imported from the same package by two other packages will have different names and so we may need some other stable ordering for types.}
+
+*Open question*: As described, this requires a pointer dereference do disambiguate pointer types.
+It would be possible to use any space in the low three bits that is not used to disambiguate non-pointer types to differentiate a small number of pointee types.
+This would save a load in pattern matching and might be useful, though given that the main reason for pattern matching is to then call a method on the concrete type, it's not clear that this is actually a win.
+Note that, in the CHERI case, the low bits are still available and so it may be useful to use them to differentiate between pointee types.
+
+These rules are somewhat complex in the abstract.
```

Comment:
What the document hasn't clarified for me, so far, is if those rules are static (ie. the same for any union type, however complex) or will be optimised to allow corner cases to be represented in the discriminator bits IFF the compiler can guarantee those union types won't intersect with the common case in code.

While I think that kind of optimisation is doable, I believe it makes the life of the compiler more complicated and by consequence helps introduce bugs in code generation that are hard to detect, debug and create tests for.

Reply:
So, doing the same lowering for any (canonicalised) union type can mean two things:
1. Each case has its own representation, so `(U8 | U16)` can have a different bit-pattern to represent `U8` than in `(U8 | U32)`, but all `(U8 | U32)` will have the same bit-pattern for `U8` and `U32`.
2. Each type is represented in the same way, regardless of the union type, so `U8` would always be the same bit-pattern, no matter the union it's in.

(1) allows us to (consistently) represent a larger set of union types (including classes, interfaces, singletons, pointers), but also requires us to know how to read the discriminators from each possible union. This can be done with a cached version of the patterns on a map to avoid recomputation, but it can also lead to less clear code-gen, which is harder to debug.

(2) allows us to have very predictable code-gen, but at the expense that any non-primitive type will require more loads and branches than they could.

I don't like (2)'s run-time costs, but I also don't like (1)'s compile time costs. I guess we can't just pick one now and change later without some major refactoring on the implementation of the compiler as well as some builtin behaviour.

I'm not sure what is best, tbh, and I'm happy with either if people feel strongly about them.

Reply:
I don't think the compile-time cost for calculating the layouts will be very high (and it can be memoised).  

My bias is towards different representations for each union type and explicit operations to transform, but I am also aware of the objection that we may end up with a lot of those that are not no-op casts and that can dominate performance.  I would like it if the compiler could:

 - Preserve these as opaque things for as long as possible.
 - Have an early and at-lowering-to-LLVM point for eliminating no-op casts.
 - Initially use a very simple model for these.

Once we have a reasonable corpus, we can extract the sets of types and casts from the compiler and I can spend some time with Z3 trying to tune the set of representations to minimise the number of conversions that require anything other than a no-op.

Reply:
This is one possible candidate for a dialect operation, then. That's the only sane way to keep semantics around without pattern-matching later, and we can perform the actual computation (and elision) at the "lower-to-LLVM-dialect" stage.